napari/components/dims.py

```diff
@@ -1,265 +1,220 @@
+                ndim - len(self.current_step)
+            ) + self.current_step
+        elif len(self.current_step) > ndim:
+            self._current_step = self.current_step[-ndim:]
+
+        if len(self.order) < ndim:
+            self._order = tuple(range(ndim - len(self.order))) + tuple(
+                o + ndim - len(self.order) for o in self.order
             )
-        ]
-        return point
+        elif len(self.order) > ndim:
+            self._order = reorder_after_dim_reduction(self.order[-ndim:])
 
-    @property
-    def axis_labels(self):
-        """List of labels for each axis."""
-        return copy(self._axis_labels)
+        if len(self.axis_labels) < ndim:
+            # Append new "default" labels to existing ones
+            if self.axis_labels == tuple(
+                map(str, range(len(self.axis_labels)))
+            ):
+                self._axis_labels = tuple(map(str, range(ndim)))
+            else:
+                self._axis_labels = (
+                    tuple(map(str, range(ndim - len(self.axis_labels))))
+                    + self.axis_labels
+                )
+        elif len(self.axis_labels) > ndim:
+            self._axis_labels = self.axis_labels[-ndim:]
 
-    @axis_labels.setter
-    def axis_labels(self, labels):
-        if self._axis_labels == labels:
-            return
+        self._ndim = ndim
 
-        if len(labels) != self.ndim:
+    def _on_order_set(self, order):
+        if not set(order) == set(range(self.ndim)):
+            print('eeeeee')
             raise ValueError(
-                f"Number of labels doesn't match number of dimensions. Number"
-                f" of given labels was {len(labels)}, number of dimensions is"
-                f" {self.ndim}. Note: If you wish to keep some of the "
-                "dimensions unlabeled, use '' instead."
+                f"Invalid ordering {order} for {self.ndim} dimensions"
             )
 
-        self._axis_labels = list(labels)
-        for axis in range(self.ndim):
-            self.events.axis_labels(axis=axis)
-
-    @property
-    def last_used(self):
-        """int: Index of the last used slider."""
-        return self._last_used
-
-    @last_used.setter
-    def last_used(self, last_used):
-        if self._last_used == last_used:
-            return
-        self._last_used = last_used
-        self.events.last_used()
-
-    @property
-    def order(self):
-        """List of int: Display order of dimensions."""
-        return copy(self._order)
-
-    @order.setter
-    def order(self, order):
-        if np.all(self._order == order):
-            return
-
-        if not len(order) == self.ndim:
+    def _on_axis_labels_set(self, axis_labels):
+        if not len(axis_labels) == self.ndim:
             raise ValueError(
-                f"Invalid ordering {order} for {self.ndim} dimensions"
+                f"Invalid number of axis labels {len(axis_labels)} for {self.ndim} dimensions"
             )
```

Comment:
I see that we heck here ndim, but in code of ndim there is no interaction with axis labels.

Reply:
The axis_labels length are adjusted inside the `ndim` set method here https://github.com/napari/napari/blob/e4c7b37a7c79602c8b1ea7d543bb3a8cfdbe1f14/napari/components/dims.py#L162-L174. Is that what you were looking for?

Reply:
Yes. sorry.